Heimdallr/OAuthAuthorizationCodeHandler.swift

```diff
@@ -0,0 +1,142 @@
+import Foundation
+import UIKit
+import SafariServices
+import Result
+
+/// The request could not be authorized (e.g., no refresh token available).
+public let HeimdallrErrorAuthorizationCanceled = 101
+
+private enum CompletionType {
+    case authorizationCode(AuthorizationCodeCompletion)
+    case accessToken(AccessTokenCompletion)
+    case none
+}
+
+extension CompletionType {
+
+    func oAuthAccessToken(fromURL url: URL) -> OAuthAccessToken? {
+
+        guard let paramters = url.fragmentParameters,
+            let token = paramters["access_token"] else { return nil }
+
+        return OAuthAccessToken(accessToken: token)
+    }
+
+    func handle(url: URL) {
+        switch self {
+        case let .authorizationCode(completion):
+            guard let authorizationCode = url.queryParameters?["code"] else {
+                completion(.failure(NSError(domain: "", code: 0, userInfo: nil)))
+                return
+            }
+            completion(.success(authorizationCode))
+        case let .accessToken(completion):
+            guard let oAuthAccessToken = oAuthAccessToken(fromURL: url) else {
+                completion(.failure(NSError(domain: "", code: 0, userInfo: nil)))
+                return
+            }
+            completion(.success(oAuthAccessToken))
+        case .none:
+            break
+        }
+    }
+
+    func fail(error: NSError) {
+        switch self {
+        case let .authorizationCode(completion):
+            completion(.failure(error))
+        case let .accessToken(completion):
+            completion(.failure(error))
+        case .none:
+            assertionFailure("Undefined state")
+            break
+        }
+    }
+}
+
+protocol OAuthAuthorizationCodeHandlerType {
+
+    func requestAuthorizationCode(url: URL, completion: @escaping AuthorizationCodeCompletion)
+    func requestAccessToken(url: URL, completion: @escaping (Result<OAuthAccessToken, NSError>) -> Void)
+    func authCallback(url: URL?, error: Error?)
+}
+
+@available(iOS 9.0, *)
+class OAuthAuthorizationCodeHandler: NSObject, OAuthAuthorizationCodeHandlerType {
+
+    private var authenticationSession: Any?
+    private var safariViewController: SFSafariViewController?
+    private var completion: CompletionType = .none
+
+    func requestAuthorizationCode(url: URL, completion: @escaping AuthorizationCodeCompletion) {
+        self.completion = .authorizationCode(completion)
+
+        if #available(iOS 11.0, *) {
+            requestWithSafariServices(url: url)
+        } else {
+            requestLegacy(url: url)
+        }
```

Comment:
How about having one function that handles the URL that internally checks for availability of iOS 11? That way we don't have to do the check twice (here and in `requestAccessToken`)?